### ISSUE TITLE
cast create2 wrong address/salt calculation

### ISSUE THREAD

**Issue by karmacoma-eth** [foundry-rs/foundry#3510]
### Component

Cast

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

`forge 0.2.0 (fb4a836 2022-10-18T00:12:04.315108Z)`

### What command(s) is the bug in?

cast create2

### Operating System

macOS (Intel)

### Describe the bug

I am not discounting the possibility that I'm making a mistake but I'm not able to reproduce the salt/address combinations reported by `cast create2`.

Here is a script you can run to demonstrate the difference:

```
#!/bin/sh

set -e

INIT_CODE_HASH=0x8117a76146b52c01f79ae3b7f0e04ef9ac84a00974c0144222ded76ab3dadcf3
DEPLOYER=0x4e59b44847b379578588920ca78fbf26c0b4956c

function computeAddress() {
    local INIT=$1
    salt=$(cast create2 \
        --init-code $INIT \
        --ends-with dead \
        --deployer $DEPLOYER \
        | tee create2.log | grep Salt | cut -d' ' -f2)

    hex_salt=$(cast --to-base $salt 16)
    hex_data=$(cast --concat-hex 0xff $DEPLOYER $hex_salt $INIT_CODE_HASH)
    hashed_data=$(cast keccak $hex_data)

    echo 'using salt' $hex_salt
    echo 'expected addr:' $(grep 'Address' create2.log | cut -d' ' -f2)
    echo 'computed addr:' $(echo $hashed_data | python3 -c 'import sys; print(f"0x{sys.stdin.read()[-41:]}")')
}

echo 'using cast create2 --init-code $INIT_CODE_HASH'
computeAddress $INIT_CODE_HASH
```

I get the following output:

```
using cast create2 --init-code $INIT_CODE_HASH
using salt 0x202d80facd2c9fc71b201f7548f4cb371f413d841941924a2a05375de64861d3
expected addr: 0xce64Cef6d988fd714714835949edc483055Bdead
computed addr: 0xa766d80f7a8f85954cdbe27a86af88d2bb515dc6
```

(the expected addr ends with `dead` but not the computed addr)

The computed address matches what I get when I deploy with `new Contract{salt}()` using `forge script`

**Comment by DaniPopes** [comment]
Error is in this line:

https://github.com/foundry-rs/foundry/blob/870da6f73ee6ede429ed5742bb91eed3121071e3/cli/src/cmd/cast/create2.rs#L107

`init_code` before (`String` from process args): `0x8117a76146b52c01f79ae3b7f0e04ef9ac84a00974c0144222ded76ab3dadcf3`

`init_code` after (`Bytes` after `.into_bytes()`): `0x307838313137613736313436623532633031663739616533623766306530346566396163383461303039373463303134343232326465643736616233646164636633`

CC @iFrostizz

**Comment by iFrostizz** [comment]
Thanks @DaniPopes , Bytes::from(string.as_bytes()): was silly

**Comment by iFrostizz** [comment]
@gakonst This is the implementation of the function to get the `create2` address in ethers-rs: https://docs.rs/ethers-core/0.17.0/src/ethers_core/utils/mod.rs.html#186-190
it needs the `init_code` and it keccak it inside, could we provide another method before fixing the bug to give the possibility to input either `init_code` or `init_code_hash` ?

**Comment by iFrostizz** [comment]
Could anybody open the PR again ? I provided the fixes on my fork branch.

**Comment by mattsse** [comment]
sorry, wdym by "open the PR again"?

**Comment by iFrostizz** [comment]
> sorry, wdym by "open the PR again"?

Oh, I guess that I should just open a new PR, my commits are done on the same branch: https://github.com/iFrostizz/foundry/commit/60ebb541972caeb769279d768517e589d2946cfd

**Comment by Evalir** [comment]
Closed by #3634
